### PR TITLE
feat: Ensure data read order reflects commit sequence in Iceberg tables

### PR DESCRIPTION
**Pull request by malhotrashivam**
**Before this change:**
Data files were read in the order they appeared in the snapshot, which sometimes led to new data appearing at the top of the table instead of at the bottom.

**This PR:**
- Fixes the ordering of data files to reflect the commit sequence, ensuring that newer data appears at the end of the table, as expected.
- Also, allows comparison between different `TableLocationKey` implementations using fully qualified class names as a falback.

### CODE REVIEW

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (malhotrashivam) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:146]: I have a lot of unit tests in #5989, so not adding here since its harder to unit test without the ability to append multiple data files in a single snapshot.

**Review by lbooker42 — COMMENTED** [review]
This looks good.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (malhotrashivam) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:111]: This works because our locationKey creation methods are called from inside a synchronized method.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (malhotrashivam) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:76]: This works because our locationKey creation happens from inside a synchronized method through a cache.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:100]: This, at a minimum, this breaks the transitivity assumption of comparable. Say, `a = null, b = 1, c = 2`, you might say:


* a == b is true
* a == c is true
* therefore b == c is true (oops!)

I think we'll need to pick a direction; either nulls come first, or nulls come last.

In that vein, we can probably use primitives and set them to `Long.MIN_VALUE` or `Long.MAX_VALUE` in the constructor and use a normal primitive comparison.

**Review by devinrsmith — CHANGES_REQUESTED** [review]

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:100]: I would prefer keeping them at the end then

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:92]: Usually when a comparator is changed like this, we need to ensure that equals (and to a lesser extend hashcode) are updated to be consistent.

I see that the super class just uses uri and partitions for equals and hashcode; I think _normally_ this would be good enough, but I wonder if it's ever possible to use the same uri (and in the worst case, the comparisons above are all equal or all null); in which case, maybe we need to actually think use ManifestFile as part of equality / comparison logic?

I think we may need to consider a parquet location key as ManifestFile + DataFile?

Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:92]: So ManifestFile provides an additional 
```ManifestFile::sequenceNumber() : Returns the sequence number of the commit that added the manifest file.```

I can add that to the mix, after URI.

**Review by devinrsmith — DISMISSED** [review]

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:154]: I think we should drop this. It likely adds little-to-no benefit (or harm), but it might lead downstream code to be complacent and assume ordering where it should not.
```suggestion
```

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:68]: Maybe expand this to mention manifest file? Could use `@param`...

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:24]: I would use a regular comment, or document all the fields individually to explain what they are (might be better).

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:65]: I don't understand why this is here. If it's for the case where `dataSequenceNumber` and `fileSequenceNumber` are both `null`, my reading suggests that we're not going to get any additional correctness from looking at `manifestSequenceNumber`.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:81]: Out of date.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:106]: These should all be *after* partitions, but before uri, unless we're deciding that these are the actual determiners of order. Historically, we've wanted to group related data via partitions.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:128]: I actually think this might be a bad change. We need to decide how adding a URI at different sequence numbers should be handled.
I could see this creating a spurious remove and add pair.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:133]: We could probably hash the uri first...

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:72]: So `dataSequenceNumber` and `fileSequenceNumber` can be `null` for old iceberg versions in some cases.
But `manifestSequenceNumber` will never be null for such files too because we always read from some non-null manifest file.
So `manifestSequenceNumber + dataFilePos` will give us some way to order such files.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:141]: Even though we the devs are saying uri is guaranteed to be unique, we've seen that we may not be able to rely on that. As such, I think we need to be more defensive at this lower layer. Higher layers can choose to disallow duplicated URIs, but I don't think as the key we can.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:143]: Keeping it this way since we saw cases where different data files had the same URIs.
Added a comment explaining the same.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUpdateMode.java
Comment (lbooker42) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUpdateMode.java:41]: We should add the data ordering caveat to the other `autoRefreshingMode()` function as well.

**Review by devinrsmith — APPROVED** [review]
Please update the description / merge message for to match the latest behavior before merging.

**Review by rcaudy — COMMENTED** [review]
.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:167]: Oops...

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:122]: This is a good first step, but I'm realizing we have some trouble here.
We may have neglected this, but TLKs really must be comparable to all TLKs, not just TLKs of the same provenance.
We need fallback comparisons.

We have two paths forward:
1. Keep behavior as-is, and defer to a new issue:  Go back to throwing an exception on this line for now. Create an issue to define an ordering.
2. Fix the problem now.

My proposal for fixing, whether we do it here or in a new issue:
1. Make unrelated implementations compare to one another based on fully-qualified class name. This imposes an arbitrary but consistent ordering, and should roughly keep distinct sources together.
2. Make PartitionedTLK use order + partitions, followed by classname for tie-breaking.
3. Make URITLK order + partitions, then break ties by URI (if the other is a URITLK), and lastly break ties by classname.
4. Make IcebergTLK compare to other IcbergTLKs by Catalog name and TableIdentifier (which should also be added to equals and hashcode), then as currently defined in this PR, then to other URITLKs as defined by the URITLK rules, and so on.

Might want opinions from @abaranec and @devinrsmith

**Review by rcaudy — COMMENTED** [review]
.

File: engine/test-utils/src/main/java/io/deephaven/engine/testutil/locations/TableBackedTableLocationKey.java
Comment (rcaudy) [inline engine/test-utils/src/main/java/io/deephaven/engine/testutil/locations/TableBackedTableLocationKey.java:52]:
```suggestion
            final int idComparisonResult =
```

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:127]: This is neglecting the namespace part of a `TableIdentifier`. Maybe you want `toString()` instead of `name()`?

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java
Comment (rcaudy) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergBaseLayout.java:100]: That sure is a lot of constructor args. Sorry. Builder? (Probably not worth it for something called in only once place.)

**Review by rcaudy — DISMISSED** [review]
I'm happy if @abaranec and @devinrsmith are.

**Review by devinrsmith — CHANGES_REQUESTED** [review]

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java
Comment (rcaudy) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/location/IcebergTableParquetLocationKey.java:152]: That's precisely why I think it should be included. If the user were to create a TLP sourcing from two distinctly named catalogs, I'd want them to be able to include an otherwise-identical location from each. Maybe it's too hypothetical.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java:13]: Could be package private.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java:15]: This could be singleton enum, no need to construct.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java
Comment (devinrsmith) [inline extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java:40]: I would formulate this with the `!= 0` style, reads more linearly.

**Review by rcaudy — COMMENTED** [review]
.

File: extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java
Comment (devinrsmith) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java:13]: Okay, move the comparator to somewhere else then. Could be top level package private.

Comment (malhotrashivam) [reply extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergUtils.java:13]: I made the comparator instance public and comparator itself private, since we need to access the instance from IcebergTableParquetLocationKey. Let me know if this is better.

**Review by rcaudy — DISMISSED** [review]

**Review by rcaudy — APPROVED** [review]

**Review by devinrsmith — APPROVED** [review]

**Pull request merged**
